File: week1/pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int get_lenght(const long n);
bool luhns_algorithm_passed(const long n, int len);
bool print_card_type(const long number, const int lenght);

int main(void)
{
    long credit_card_number = get_long("Number: ");
    int number_lenght = get_lenght(credit_card_number);
    if (number_lenght != 0 && luhns_algorithm_passed(credit_card_number, number_lenght))
    {
        print_card_type(credit_card_number, number_lenght); //return true if credit card number is invalid
    }
    else
    {
        printf("INVALID\n");
    }
}

int get_lenght(const long n)  // * ANY 0s IN FRONT OF n WILL NOT BE TAKEN INTO ACCOUNT
//the returning value 0 will mean the credit number is not valid.
{
    if (n <= 0) //doesn't allow any n <= 0.
    {
        return 0;
    }
//give the lenght of the number; e. g. to get the lenght of 1000:
//log(1000)/log(10) + 1 = log10(1000) + 1 = 3 + 1 = 4.
    int l = floor(log(n) / log(10) + 1); // * the operation ignores any 0s in front of n!
    if (l > 16)
    {
        return 0;
    }
    else
    {
        return l;
    }
}

bool luhns_algorithm_passed(const long n, int len)
{
    int sum = 0;
    long z = n;
    for (int i = len; i > 0; i --)
    {
        int last_digit = z % 10;
        if ((len - i) % 2 == 0) //add the last digit to sum.
        {
            sum += last_digit;
        }
        else //add the sum of the digits of the double of the last digit.
        {
            if (last_digit < 5)
            {
                sum += 2 * last_digit;
            }
            else //adds app the digits of a number if its double was composed by two digits.
            {
                sum += floor((last_digit * 2) / 10) + (last_digit * 2) % 10;
            }
        }
        z = floor(z / 10); //removes the last digit from n.
    }
    //check if card does respect luhn's algorithm
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool print_card_type(const long number, const int lenght) //returns true if card number is invalid
{
    //get first two digits:
    int first_two_dig = floor(number / pow(10, (lenght - 2)));
    //get first digit:
    int first_dig = floor(number / pow(10, lenght - 1));

    if (lenght == 15)
    {
        if (first_two_dig == 34 || first_two_dig == 37)
        {
            printf("AMEX\n");
            return false;
        }
        printf("INVALID\n");
        return true;
    }
    else if (lenght == 13 && first_dig == 4)
    {
        printf("VISA\n");
        return false;
    }
    else if (lenght == 16)
    {
        if (first_dig == 4)
        {
            printf("VISA\n");
            return false;
        }
        else if (50 < first_two_dig && first_two_dig < 56)
        {
            printf("MASTERCARD\n");
            return false;
        }
        printf("INVALID\n");
        return true;
    }
    else
    {
        printf("INVALID\n");
        return true;
    }
}

File: week6/pset6/sentimental-credit/credit.py
from cs50 import get_int
import math
import sys


def main():
    number = get_int("Number: ")

    if number < 0:
        print("INVALID")
        sys.exit(-1)

    if not luhns_algorithm(str(number)):
        print("INVALID")
        sys.exit(-2)

    ''' Print cart type '''
    c_type = card_type(str(number))
    print(c_type)


def luhns_algorithm(num):  # num is string-type
    sum = 0
    for i in range(1, len(num) + 1):
        if i % 2 != 0:
            sum += int(num[-i])
        else:
            d = int(num[-i]) * 2
            if d < 10:
                sum += d
            else:
                ''' add the sum of the two digits of d to sum '''
                sum += math.floor(d / 10) + (d % 10)
    if sum % 10 == 0:
        return True
    return False


def card_type(num):  # num is string-type
    if len(num) == 15:
        if num[0:2] == "34" or num[0:2] == "37":
            return "AMEX"
        return "INVALID"
    elif len(num) == 13 and num[0] == "4":
        return "VISA"
    elif len(num) == 16:
        if num[0] == "4":
            return "VISA"
        elif 50 < int(num[0:2]) and int(num[0:2]) < 56:
            return "MASTERCARD"
        return "INVALID"
    return "INVALID"


if __name__ == "__main__":
    main()
    sys.exit(0)
